### PR TITLE
[MM-12068] Add ability to remove custom branding image

### PR DESCRIPTION
**Pull request by hahmadia**
**NOTE**
I am not very experienced and understanding of how the styling works within the APP. If someone could explain how to style a className that I attach to a `<div/>`, I would greatly appreciate it because I couldn't figure it out. Other than the styling, I think all else should be okay (unless I have implemented something incorrectly)

As a junior dev and a first contribution, all and any advice would be greatly appreciated. 
Thank you!

#### Summary

This ticket gives the user the ability to remove custom branding image

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-server/issues/9507

### Demonstration Video
MM-12068

### CODE REVIEW

**Comment by hanzei** [comment]
Hey @hahmadia,

looks like you have a linter error:
```
eslint --ignore-pattern node_modules --ignore-pattern non_npm_dependencies --ignore-pattern dist --ext .js --ext .jsx . --quiet


/root/mattermost-webapp/components/admin_console/schema_admin_settings.jsx
  600:41  error  use `Boolean(saveNeeded)` instead  no-implicit-coercion
```

**Comment by hahmadia** [comment]
Thanks for pointing that out @hanzei. I have fixed all lint + translation issues.

**Comment by esethna** [comment]
Thanks for the PR @hahmadia! @asaadmahmood may be able to help provide some styling guidance

**Comment by asaadmahmood** [comment]
One thing that's not working is if I lets say I don't have any image uploaded, but if I do upload it, the save button does not turn blue, I have to change some other setting to make it recognise that something has changed.

**Comment by asaadmahmood** [comment]
@hahmadia I have updated the css, please pull my changes from your remote branch before making any changes locally and pushing them.

**Comment by asaadmahmood** [comment]
@esethna We have an outstanding question on UX, which is, whether the save button should detect a change when a user uploads a picture.

The way it's currently implemented is that if a user uploads a picture, the save button does not change because I think it automatically uploads and saves it without the user performing a "Save" action like he does for all the other fields.

**Comment by esethna** [comment]
Thanks @hahmadia @asaadmahmood, as per the designs the user should have to hit the "Save" button after selecting an image from the file window:

image

**Comment by hahmadia** [comment]
@asaadmahmood @esethna 
Thank you for the feedback! I have addressed the issues. I had a misunderstanding with how the upload process should work.

**Comment by hmhealey** [comment]
@esethna @asaadmahmood If it didn't work that way before, I think that should be turned into a separate ticket. Tying the file upload into the save button is going to be more difficult, and I don't think we need to block this feature because of that.

**Comment by wiersgallak** [comment]
@hahmadia When removing the custom brand image, I would expect the login page to center as it does from default when there is no custom brand image.  Currently, after removing the custom brand image, the login page stays formatted to the right as if the image was still there. 

Screen Shot 2019-08-06 at 8 50 52 AM

**Comment by hmhealey** [comment]
@wiersgallak That's existing behaviour. Even if no image was uploaded in the first place, it leaves that gap as long as custom branding is enabled

File: components/admin_console/brand_image_setting/brand_image_setting.jsx [hunk 1]
```diff
@@ -62,86 +70,129 @@ export default class BrandImageSetting extends React.PureComponent {
...
+        if (this.props.saving && this.state.deleteBrandImage && !this.props.disabled) {
+            this.handleImageDelete();
+        }
     }
 
     handleImageChange() {
         const element = $(this.refs.fileInput);
-
         if (element.prop('files').length > 0) {
+            this.props.onChange('saveNeeded', 'config');
```
Comment (hmhealey) [inline components/admin_console/brand_image_setting/brand_image_setting.jsx]: Maybe we could pass in a separate `setSaveNeeded` prop to use for this instead of using `onChange`? This works, but it isn't really how `onChange` was intended to work, so it could get broken accidentally in the future.

Alternatively, if we end up going with something like `registerSaveAction` like I mentioned above, that could set `saveNeeded` for you when you call it.

File: components/admin_console/brand_image_setting/brand_image_setting.jsx [hunk 2]
```diff
@@ -62,86 +70,129 @@ export default class BrandImageSetting extends React.PureComponent {
...
+                                id='admin.team.removeBrandImage'
+                                defaultMessage='Remove brand image'
+                            />
+                        </div>
+                    </Tooltip>
+                )}
+            >
+                <button
+                    className='remove-image__btn'
+                    onClick={() => {
```
Comment (hmhealey) [inline components/admin_console/brand_image_setting/brand_image_setting.jsx]: Could you make this a method instead of using an arrow function? We generally try to avoid using arrow functions in `render` because it can cause child components to re-render when they don't need to.

File: components/admin_console/brand_image_setting/brand_image_setting.jsx [hunk 3]
```diff
@@ -62,86 +70,129 @@ export default class BrandImageSetting extends React.PureComponent {
...
-        }
-
         let letbtnDefaultClass = 'btn';
         if (!this.props.disabled) {
             letbtnDefaultClass += ' btn-default';
         }
 
         let img = null;
+        let displayImgOverlay = false;
+        const IMGOVERLAY = (
```
Comment (hmhealey) [inline components/admin_console/brand_image_setting/brand_image_setting.jsx]: Can you use camel case for this name since it's not a constant?

File: components/admin_console/brand_image_setting/brand_image_setting.jsx [hunk 4]
```diff
@@ -62,86 +70,129 @@ export default class BrandImageSetting extends React.PureComponent {
...
-                    src={Client4.getBrandImageUrl(this.state.brandImageTimestamp)}
-                />
+                <div className='remove-image__img margin-bottom x3'>
+                    <img
+                        alt='brand image'
+                        src={Client4.getBrandImageUrl(this.state.brandImageTimestamp)}
+                    />
+                </div>
             );
+            displayImgOverlay = !this.props.disabled;
```
Comment (hmhealey) [inline components/admin_console/brand_image_setting/brand_image_setting.jsx]: Since we only use `IMGOVERLAY` in this one case, maybe we could move it so that we only define it here? We could do something like
```javascript
} else if (this.state.brandImageExists) {
    let overlay;
    if (!this.props.disabled) {
        overlay = (...);
    }

    img = (
        <div className='remove-image__img margin-bottom x3'>
            <img .../>
            {overlay}
        </div>
    );
} else {
```

File: components/admin_console/schema_admin_settings.jsx [hunk 5]
```diff
@@ -812,10 +816,6 @@ export default class SchemaAdminSettings extends React.Component {
                 if (callback) {
                     callback();
                 }
-
-                if (this.handleSaved) {
```
Comment (hmhealey) [inline components/admin_console/schema_admin_settings.jsx]: Why remove this?

File: components/admin_console/brand_image_setting/brand_image_setting.jsx [hunk 6]
```diff
@@ -62,86 +70,129 @@ export default class BrandImageSetting extends React.PureComponent {
 
             reader.readAsDataURL(this.state.brandImage);
         }
+
+        if (this.props.saving && this.state.brandImage && !this.props.disabled) {
+            this.handleImageSubmit();
```
[the remark this answers is not in the document]
Comment (jespino) [reply components/admin_console/brand_image_setting/brand_image_setting.jsx]: Yes, I think here, the "cleanest" approach is to use refs. I you reference a node which is a component, you get the instance of the component and you can call the internal methods of the component itself, so, I would add a ref in the parent to the instance of this component and then call. this.brandImageSettingRef.current.saveImage().

The callback to register a callback can work, but sounds complicated.

Use methods of a component from the outside is rare and normally unexpected for the developers because is not the common use case, so I would suggest to add a comment in the function like `// This function is used from the outside to trigger the save of the image when needed.`

File: components/admin_console/schema_admin_settings.jsx (hunk 5 above)
Comment (hahmadia) [reply components/admin_console/schema_admin_settings.jsx]: The reason I removed this was I couldn't find any corresponding any handleSaved function in the file and thought maybe this was added by accident so I decided to remove it. I just wasn't quite sure why we had it and where it's corresponding function call is. I'll add it back in but would just like to know where the corresponding handleSaved function is. Thanks!

File: components/admin_console/brand_image_setting/brand_image_setting.jsx (hunk 1 above)
Comment (hahmadia) [reply components/admin_console/brand_image_setting/brand_image_setting.jsx]: Resolved!

File: components/admin_console/brand_image_setting/brand_image_setting.jsx (hunk 2 above)
Comment (hahmadia) [reply components/admin_console/brand_image_setting/brand_image_setting.jsx]: Resolved!

File: components/admin_console/brand_image_setting/brand_image_setting.jsx (hunk 3 above)
Comment (hahmadia) [reply components/admin_console/brand_image_setting/brand_image_setting.jsx]: Resolved!

File: components/admin_console/brand_image_setting/brand_image_setting.jsx (hunk 4 above)
Comment (hahmadia) [reply components/admin_console/brand_image_setting/brand_image_setting.jsx]: Resolved!

File: components/admin_console/schema_admin_settings.jsx (hunk 5 above)
Comment (hahmadia) [reply components/admin_console/schema_admin_settings.jsx]: I've added this back in. I'm just not sure why we have it there or what the purpose is? Thanks!

**Comment by hahmadia** [comment]
@hmhealey @jespino 
Thank you for all your suggestions! I have addressed them.
I was totally not aware about the createRef. Although I had seen it, I wasn't exactly sure what its purpose was so I'm glad it was mentioned as it allowed me to research it and understand it better. Hope to get a re-review and address any other issues that you guys may come across.

File: components/admin_console/schema_admin_settings.jsx [hunk 7]
```diff
@@ -586,12 +588,11 @@ export default class SchemaAdminSettings extends React.Component {
         if (this.state.saveNeeded === 'permissions') {
             saveNeeded = 'both';
         }
+
         this.setState({
             saveNeeded,
             [id]: value,
         });
-
-        this.props.setNavigationBlocked(true);
```
Comment (hahmadia) [inline components/admin_console/schema_admin_settings.jsx]: Note to self: Add this back in.

File: components/admin_console/brand_image_setting/brand_image_setting.jsx [hunk 8]
```diff
@@ -22,20 +21,30 @@ export default class BrandImageSetting extends React.PureComponent {
...
+
+        /*
+        * Enable save button when image is deleted
+        */
+        setSaveNeeded: PropTypes.func.isRequired,
+
+        /*
+        * Is the component saving or not
+        */
+        saving: PropTypes.bool.isRequired,
```
Comment (hahmadia) [inline components/admin_console/brand_image_setting/brand_image_setting.jsx]: Note to self: Delete saving and reword comment for setSaveNeeded

File: components/admin_console/schema_admin_settings.jsx [hunk 9]
```diff
@@ -787,7 +797,10 @@ export default class SchemaAdminSettings extends React.Component {
         this.setState({errorTooltip: isElipsis});
     }
 
-    doSubmit = (callback, getStateFromConfig) => {
+    doSubmit = async (callback, getStateFromConfig) => {
```
Comment (jespino) [inline components/admin_console/schema_admin_settings.jsx]: The changes here are converting this "generic" approach into a specific case handling approach. I'm not against the idea, but I'm against make it specific to the BrandImage. By the way, you are reusing the same reference or multiple custom element which is going to break if you have other custom element in the page after this one.

So, you have 2 problems here, 1 reference that is used for multiple components (that shouldn't happen), and one handling everything specifically for the "CustomBrandImage" component, which shouldn't be.

My proposal, if we want to follow this path, is to have an array of references, and for each reference try to run the an handleSave function.

something like:
```
const customComponent = this.customComponentRef.current;
if (customComponent.handleSave) {
    result = await customComponent.handleSave()
}
```

Other important thing, you are accessing to the "local state" of the component from the outside, this is allowed by the references, but because you can don't mean you should. The local state should be maintained private, so I would restrict the usage of the references to call the concrete methods of the components.

Comment (hahmadia) [reply components/admin_console/schema_admin_settings.jsx]: @jespino Thank you so much for pointing these 2 problems out. I hadn't thought of referencing multiple components being an issue but I did have the issue of making a generic function specific to the `CustomBrandImage` component. I've addressed both of these problems by making it more general and having an array of references. 

I'll also be sure to keep in mind not to access state from outside the component too!
Thanks for your feedback and help 😃

Comment (hmhealey) [reply components/admin_console/schema_admin_settings.jsx]: I don't know if we can keep track of multiple refs by pushing them onto an array. When using a ref callback like this, it gets called with `null` when the child component is unmounted. It's easy enough to avoid adding a null ref to that array, but I'm wondering if we might run into any problems because we have no way to remove entries from that array.

I'd also be worried about using refs here because they can get a bit tricky with higher order components like ones created using `connect(MyComponent)` to connect to the redux state, but that's just a problem with refs in general.

@jespino What's the motivation for preferring refs over something like I suggested where the `SchemaAdminSettings` passes a method into the child component that the child can use to tell the `SchemaAdminSettings` that it has something to save? I'm 0/5 either way, but that might be easier than using refs

Comment (jespino) [reply components/admin_console/schema_admin_settings.jsx]: The problem with that approach is how hard to follow is it (passing a function that allow to set the function used to save from the parent), but yes, the "array of refs" looks like complicated and more error prone. I'm 2/5 here, but I guess your solution, for this case is not perfect, but is less error prone so I think I prefer it.

Comment (hahmadia) [reply components/admin_console/schema_admin_settings.jsx]: Requested changes have been made 🎉

File: components/admin_console/brand_image_setting/brand_image_setting.jsx [hunk 10]
```diff
@@ -66,56 +79,65 @@ export default class BrandImageSetting extends React.PureComponent {
...
+                    });
+                },
+                (err) => {
+                    this.setState({
+                        error: err.message,
+                    });
+                }
+            );
+        }
+        return this.state.error;
```
Comment (hmhealey) [inline components/admin_console/brand_image_setting/brand_image_setting.jsx]: I don't think this will return an error like you're expecting. `this.setState` doesn't update `this.state` immediately, and in this case, I think it doesn't happen until after we're done responding to the click event that started the submit process.

It would be a bit easier to structure this if it wasn't using these callback-based actions, but you could do something like this to get the error returned:
```javascript
let errorMessage;
if (this.state.deleteImage) {
    await deleteBrandImage(
        () => {...},
        (err) => {
            errorMessage = err.message;
        }
    );
}

return errorMessage;
```

Comment (hmhealey) [reply components/admin_console/brand_image_setting/brand_image_setting.jsx]: Also, it might be better to return an object like `{error: ...}` to indicate that an error occurred since that's more consistent with how we tend to return errors from other async functions

File: components/admin_console/schema_admin_settings.jsx [hunk 11]
```diff
@@ -833,6 +838,23 @@ export default class SchemaAdminSettings extends React.Component {
...
         );
+
+        const results = [];
+        for (const saveAction of this.saveActions) {
+            results.push(saveAction());
+        }
+
+        // This waits for all the save actions to be complete and then checks to see if any of the values are truthy value to determine
+        // if they failed or not (since they should return either null or '' on a successfull call and some error text on a failed operation)
+        const hasSaveActionError = await Promise.all(results).then((values) => values.some(((value) => value))).then((value) => value);
```
Comment (hmhealey) [inline components/admin_console/schema_admin_settings.jsx]: I don't think that last `.then` should be necessary here
```suggestion
        const hasSaveActionError = await Promise.all(results).then((values) => values.some(((value) => value)));
```

Also, if you change the return type of the save action, you'll need to account for that here

File: components/admin_console/schema_admin_settings.jsx [hunk 12]
```diff
@@ -833,6 +838,23 @@ export default class SchemaAdminSettings extends React.Component {
...
+
+        const hasError = this.state.serverError || hasSaveActionError;
```
Comment (hmhealey) [inline components/admin_console/schema_admin_settings.jsx]: We should probably set `this.state.serverError` if the save action had an error

Comment (hahmadia) [reply components/admin_console/schema_admin_settings.jsx]: As discussed through direct communication, `BrandImageSettings` displays its own error so this is not required.

File: components/admin_console/schema_admin_settings.jsx (hunk 11 above)
Comment (hahmadia) [reply components/admin_console/schema_admin_settings.jsx]: Thanks for the catch! I've addressed this

File: components/admin_console/brand_image_setting/brand_image_setting.jsx (hunk 10 above)
Comment (hahmadia) [reply components/admin_console/brand_image_setting/brand_image_setting.jsx]: Thanks! I've now addressed this.

**Comment by hahmadia** [comment]
@jespino I have addressed your concern regarding multiple actions being saved if the component gets mounted and unmounted. Thanks!

**Comment by jespino** [comment]
@hahmadia On thing, you forced a push to the PR, we prefer to not do that because that means for us the need to re-review everything before we approve again. If you want to synchronize with master just merge master. At the end we going to "Squash and merge" everything in one single commit, so the history is going to be clean anyway.

**Comment by hahmadia** [comment]
@jespino Alright thanks for the advice. I will be sure to do this for my future changes.

File: components/admin_console/brand_image_setting/brand_image_setting.jsx [hunk 13]
```diff
@@ -22,20 +21,35 @@ export default class BrandImageSetting extends React.PureComponent {
...
+        /*
+        * Unregisters the function on unmount of the component suppose to be run when the save button is pressed
+        */
+        unRegisterSaveAction: PropTypes.func.isRequired,
     }
 
     constructor(props) {
         super(props);
 
         this.handleImageChange = this.handleImageChange.bind(this);
```
Comment (jespino) [inline components/admin_console/brand_image_setting/brand_image_setting.jsx]: Not a stopper, but we normally prefer to use "arrow functions" to avoid the need of binding this, for example if you define the `handleImageChange` in the component as:
```js
    handleImageChange = () => {
        ...
    }
```
instead of 
```js
    handleImageChange() {
        ...
    }
```

You wont need the binding.

File: components/admin_console/brand_image_setting/brand_image_setting.jsx [hunk 14]
```diff
@@ -181,12 +242,6 @@ export default class BrandImageSetting extends React.PureComponent {
                             onChange={this.handleImageChange}
                         />
                     </div>
-                    <UploadButton
-                        primaryClass={btnPrimaryClass}
-                        status={this.state.status}
-                        disabled={this.props.disabled || !this.state.brandImage}
-                        onClick={this.handleImageSubmit}
-                    />
```
Comment (saturninoabril) [inline components/admin_console/brand_image_setting/brand_image_setting.jsx]: Would be nice to add unit test for this component.

File: actions/admin_actions.jsx [hunk 15]
```diff
@@ -162,6 +162,15 @@ export async function uploadBrandImage(brandImage, success, error) {
...
 
+export async function deleteBrandImage(success, error) {
+    const {data, error: err} = await AdminActions.deleteBrandImage()(dispatch, getState);
+    if (data && success) {
+        success(data);
+    } else if (err && error) {
+        error({id: err.server_error_id, ...err});
+    }
+}
+
```
Comment (saturninoabril) [inline actions/admin_actions.jsx]: We're actually moving away from this pattern into a more redux-friendly format which is dispatching an action directly.  That can be done by creating ``components/admin_console/brand_image_setting/index.js`` and from there pass `uploadBrandImage` and `deleteBrandImage` as actions into ``brand_image_setting.jsx``.  That would be not necessary for this PR but good to have.
Let me know what's your plan.  That could also be done on separate PR if you're interested.

**Comment by cpanato** [comment]
will remove the test server to do some maintenance in the cloud infra. will create a new test server after that
